File: slp/backend-dotnet/Features/Llm/LlmController.cs
using System.Security.Claims;
using System.Text.Json;
using backend_dotnet.Features.Queue;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace backend_dotnet.Features.Llm;

[ApiController]
[Route("api/llm")]
[Authorize]
public class LlmController : ControllerBase
{
    private readonly ILlmService _llmService;
    private readonly ILlmLogRepository _repo;
    private readonly IQueueService _queue;
    private readonly IConfiguration _config;
    private readonly ILogger<LlmController> _logger;

    public LlmController(
        ILlmService llmService,
        ILlmLogRepository repo,
        IQueueService queue,
        IConfiguration config,
        ILogger<LlmController> logger)
    {
        _llmService = llmService;
        _repo = repo;
        _queue = queue;
        _config = config;
        _logger = logger;
    }

    // ── POST /api/llm/explain ────────────────────────────────────────────────

    [HttpPost("explain")]
    public async Task<IActionResult> Explain([FromBody] ExplainRequest request)
    {
        var userId = GetUserId();
        var prompt = _llmService.BuildExplainPrompt(request);

        if (_config.GetValue<bool>("LlmCache:Enabled", true))
        {
            var cached = await _repo.FindCachedAsync(userId, "explain", prompt);
            if (cached?.Response is not null)
            {
                _logger.LogDebug("Cache hit for explain (userId={UserId})", userId);
                return Ok(new SyncLlmResponse { Result = cached.Response });
            }
        }

        if (_config.GetValue<bool>("Queue:Enabled"))
            return await EnqueueJobAsync(userId, "explain", prompt, request);

        return await ProcessSyncAsync(userId, "explain", prompt, request);
    }

    // ── POST /api/llm/grammar-check ──────────────────────────────────────────

    [HttpPost("grammar-check")]
    public async Task<IActionResult> GrammarCheck([FromBody] GrammarCheckRequest request)
    {
        var userId = GetUserId();
        var prompt = _llmService.BuildGrammarCheckPrompt(request);

        if (_config.GetValue<bool>("LlmCache:Enabled", true))
        {
            var cached = await _repo.FindCachedAsync(userId, "grammar_check", prompt);
            if (cached?.Response is not null)
            {
                _logger.LogDebug("Cache hit for grammar_check (userId={UserId})", userId);
                return Ok(new SyncLlmResponse { Result = cached.Response });
            }
        }

        if (_config.GetValue<bool>("Queue:Enabled"))
            return await EnqueueJobAsync(userId, "grammar_check", prompt, request);

        return await ProcessSyncAsync(userId, "grammar_check", prompt, request);
    }

    // ── GET /api/llm/job/{jobId} ─────────────────────────────────────────────

    [HttpGet("job/{jobId}")]
    public async Task<IActionResult> GetJobStatus(string jobId)
    {
        var log = await _repo.GetByJobIdAsync(jobId);
        if (log is null)
            return NotFound(new { message = $"Job '{jobId}' not found." });

        var userId = GetUserId();
        if (log.UserId != userId)
            return Forbid();

        return Ok(new JobStatusResponse
        {
            JobId = log.JobId!,
            Status = log.Status ?? "Unknown",
            Result = log.Response,
            Error = log.Error,
            CreatedAt = log.CreatedAt,
            CompletedAt = log.CompletedAt
        });
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private async Task<IActionResult> ProcessSyncAsync(
        int? userId, string requestType, string prompt, object requestObj)
    {
        try
        {
            var (content, tokensUsed) = await _llmService.CallLlmAsync(prompt);

            // Persist the user-scoped log
            await _repo.CreateAsync(new LlmLog
            {
                UserId = userId,
                RequestType = requestType,
                Prompt = prompt,
                Response = content,
                TokensUsed = tokensUsed,
                Status = "Completed",
                CompletedAt = DateTime.UtcNow
            });

            // Populate / refresh global cache so this result survives LLM downtime
            await _repo.UpsertGlobalCacheAsync(requestType, prompt, content, tokensUsed);

            return Ok(new SyncLlmResponse { Result = content });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Synchronous LLM call failed for type={RequestType}", requestType);

            await _repo.CreateAsync(new LlmLog
            {
                UserId = userId,
                RequestType = requestType,
                Prompt = prompt,
                Status = "Failed",
                Error = ex.Message,
                CompletedAt = DateTime.UtcNow
            });

            return StatusCode(502, new { message = "LLM service is unavailable. Please try again later." });
        }
    }

    private async Task<IActionResult> EnqueueJobAsync(
        int? userId, string requestType, string prompt, object requestObj)
    {
        var jobId = Guid.NewGuid().ToString();

        var log = await _repo.CreateAsync(new LlmLog
        {
            UserId = userId,
            RequestType = requestType,
            Prompt = prompt,
            JobId = jobId,
            Status = "Pending"
        });

        var job = new LlmJob
        {
            JobId = jobId,
            UserId = userId,
            RequestType = requestType,
            RequestData = JsonSerializer.Serialize(requestObj),
            CreatedAt = DateTime.UtcNow,
            RetryCount = 0
        };

        await _queue.EnqueueAsync(job);

        _logger.LogInformation("Enqueued LLM job {JobId} (type={RequestType}, userId={UserId})",
            jobId, requestType, userId);

        return Accepted(new AsyncLlmResponse { JobId = jobId, Status = "Pending" });
    }

    private int? GetUserId()
    {
        var raw = User.FindFirstValue(ClaimTypes.NameIdentifier);
        return int.TryParse(raw, out var id) ? id : null;
    }
}

File: slp/backend-nestjs/src/modules/llm/llm.controller.ts
import {
  Controller,
  Post,
  Get,
  Body,
  Param,
  UseGuards,
  ForbiddenException,
  NotFoundException,
  BadRequestException,
  HttpCode,
  HttpStatus,
  Res,
  forwardRef,
  Inject,
} from "@nestjs/common";
import { Response } from "express";
import { ConfigService } from "@nestjs/config";
import { SessionGuard } from "../session/session.guard";
import { User } from "../../common/decorators/user.decorator";
import { LlmService } from "./llm.service";
import { LlmRepository } from "./llm.repository";
import { IQueueService } from "../queue/queue.service.interface";
import {
  ExplainRequest,
  GrammarCheckRequest,
  SyncLlmResponse,
  AsyncLlmResponse,
  JobStatusResponse,
} from "./llm.dto";
import { LlmJob } from "../queue/llm-job.interface";

@Controller("api/llm")
@UseGuards(SessionGuard)
export class LlmController {
  constructor(
    private llmService: LlmService,
    private llmRepo: LlmRepository,
    @Inject(forwardRef(() => IQueueService))
    private queueService: IQueueService,
    private config: ConfigService,
  ) {}

  // POST /api/llm/explain
  @Post("explain")
  async explain(@Body() request: ExplainRequest, @User() user: any) {
    const userId = user?.id ?? null;
    const prompt = this.llmService.buildExplainPrompt(request);

    // Cache check
    if (this.config.get<boolean>("LLM_CACHE_ENABLED", true)) {
      const cached = await this.llmRepo.findCachedAsync(
        userId,
        "explain",
        prompt,
      );
      if (cached?.response) {
        return { result: cached.response } as SyncLlmResponse;
      }
    }

    // Queue enabled?
    if (this.config.get<boolean>("QUEUE_ENABLED", false)) {
      return this.enqueueJob(userId, "explain", prompt, request);
    }

    // Synchronous fallback
    return this.processSync(userId, "explain", prompt);
  }

  // POST /api/llm/grammar-check
  @Post("grammar-check")
  async grammarCheck(@Body() request: GrammarCheckRequest, @User() user: any) {
    const userId = user?.id ?? null;
    const prompt = this.llmService.buildGrammarCheckPrompt(request);

    if (this.config.get<boolean>("LLM_CACHE_ENABLED", true)) {
      const cached = await this.llmRepo.findCachedAsync(
        userId,
        "grammar_check",
        prompt,
      );
      if (cached?.response) {
        return { result: cached.response } as SyncLlmResponse;
      }
    }

    if (this.config.get<boolean>("QUEUE_ENABLED", false)) {
      return this.enqueueJob(userId, "grammar_check", prompt, request);
    }

    return this.processSync(userId, "grammar_check", prompt);
  }

  // GET /api/llm/job/:jobId
  @Get("job/:jobId")
  async getJobStatus(@Param("jobId") jobId: string, @User() user: any) {
    const log = await this.llmRepo.getByJobIdAsync(jobId);
    if (!log) {
      throw new NotFoundException(`Job '${jobId}' not found`);
    }

    const userId = user?.id ?? null;
    if (log.userId !== userId) {
      throw new ForbiddenException();
    }

    const response: JobStatusResponse = {
      jobId: log.jobId!,
      status: log.status ?? "Unknown",
      result: log.response ?? undefined,
      error: log.error ?? undefined,
      createdAt: log.createdAt,
      completedAt: log.completedAt ?? undefined,
    };
    return response;
  }

  // --- Helpers --------------------------------------------------------------

  private async processSync(
    userId: number | null,
    requestType: string,
    prompt: string,
  ): Promise<SyncLlmResponse> {
    try {
      const { content, tokensUsed } =
        await this.llmService.callLlmAsync(prompt);

      // Persist user log
      await this.llmRepo.createAsync({
        userId,
        requestType,
        prompt,
        response: content,
        tokensUsed,
        status: "Completed",
        completedAt: new Date(),
      });

      // Update global cache
      await this.llmRepo.upsertGlobalCacheAsync(
        requestType,
        prompt,
        content,
        tokensUsed,
      );

      return { result: content };
    } catch (err) {
      // Log error
      await this.llmRepo.createAsync({
        userId,
        requestType,
        prompt,
        status: "Failed",
        error: err.message,
        completedAt: new Date(),
      });

      throw new BadRequestException(
        "LLM service is unavailable. Please try again later.",
      );
    }
  }

  private async enqueueJob(
    userId: number | null,
    requestType: string,
    prompt: string,
    requestObj: any,
  ): Promise<AsyncLlmResponse> {
    const jobId = this.generateJobId();

    // Create DB log with status = Pending
    const log = await this.llmRepo.createAsync({
      userId,
      requestType,
      prompt,
      jobId,
      status: "Pending",
    });

    const job: LlmJob = {
      jobId,
      userId,
      requestType,
      requestData: JSON.stringify(requestObj),
      createdAt: new Date(),
      retryCount: 0,
    };

    await this.queueService.enqueue(job);

    return { jobId, status: "Pending" };
  }

  private generateJobId(): string {
    // Simple UUID v4-like string
    return (
      Math.random().toString(36).substring(2, 15) + Date.now().toString(36)
    );
  }
}
